server/session/handler_book_edit.go

```diff
@@ -0,0 +1,75 @@
+package session
+
+import (
+	"fmt"
+	"github.com/df-mc/dragonfly/server/item"
+	"github.com/sandertv/gophertunnel/minecraft/protocol/packet"
+)
+
+// BookEditHandler handles the BookEdit packet.
+type BookEditHandler struct{}
+
+// Handle ...
+func (b BookEditHandler) Handle(p packet.Packet, s *Session) error {
+	pk := p.(*packet.BookEdit)
+
+	it, err := s.inv.Item(int(pk.InventorySlot))
+	if err != nil {
+		return fmt.Errorf("invalid inventory slot index %v", pk.InventorySlot)
+	}
+	book, ok := it.Item().(item.BookAndQuill)
+	if !ok {
+		return fmt.Errorf("inventory slot %v does not contain a writable book", pk.InventorySlot)
+	}
+
+	page := int(pk.PageNumber)
+	if page >= 50 || page < 0 {
+		return fmt.Errorf("page number %v is out of bounds", pk.PageNumber)
+	}
+	if len(pk.Text) > 256 {
+		return fmt.Errorf("text can not be longer than 256 bytes")
+	}
+
+	slot := int(pk.InventorySlot)
+	switch pk.ActionType {
+	case packet.BookActionReplacePage:
+		book = book.SetPage(page, pk.Text)
+	case packet.BookActionAddPage:
+		if len(book.Pages) >= 50 {
+			return fmt.Errorf("unable to add page beyond 50")
+		}
+		if page >= len(book.Pages) && page <= len(book.Pages)+2 {
+			book = book.SetPage(page, "")
+			break
+		}
+		if _, ok := book.Page(page); !ok {
+			return fmt.Errorf("unable to insert page at %v", pk.PageNumber)
+		}
+		book = book.InsertPage(page, pk.Text)
+	case packet.BookActionDeletePage:
+		if _, ok := book.Page(page); !ok {
+			// We break here instead of returning an error because the client can be a page or two ahead in the UI then
+			// the actual pages representation server side. The client still sends the deletion indexes.
+			break
+		}
+		book = book.DeletePage(page)
+	case packet.BookActionSwapPages:
+		if pk.SecondaryPageNumber >= 50 {
+			return fmt.Errorf("page number out of bounds")
+		}
+		_, ok := book.Page(page)
+		_, ok2 := book.Page(int(pk.SecondaryPageNumber))
+		if !ok || !ok2 {
+			// We break here instead of returning an error because the client can try to swap pages that don't exist.
+			// This happens as a result of the client being a page or two ahead in the UI then the actual pages
+			// representation server side. The client still sends the swap indexes.
```

Comment:
Would send a debug message here too.

Reply:
Same here.

Reply:
This occurs when trying to swap the last two pages while the server is not yet aware of their existence (the last two pages aren't set until input, and won't exist upon reopening in the vanilla implementation). Basically this can and will occur quite often especially since the swap button can be spammed.